weather_forecast_retrieval/data/hrrr/file_loader.py

```diff
@@ -0,0 +1,289 @@
+        if var_keys is not None:
+            var_map = {key: var_map[key] for key in var_keys}
+
+        if forecast_flag:
+            # TODO: Implement forecast retrieval here
+            raise NotImplementedError(
+                'Getting the forecast is not implemented yet')
+        else:
+            self.get_data(var_map)
+
+        metadata, dataframe = self.convert_to_dataframes(var_map)
+
+        # manipulate data in necessary ways
+        for key in dataframe.keys():
+            dataframe[key].sort_index(axis=0, inplace=True)
+            if key == 'air_temp':
+                dataframe['air_temp'] -= 273.15
+            if key == 'cloud_factor':
+                dataframe['cloud_factor'] = 1 - dataframe['cloud_factor'] / 100
+
+        return metadata, dataframe
+
+    def get_data(self, var_map):
+        """
+        Get the HRRR data for set start and end date.
+        Read data is stored on instance attribute.
+
+        hours    0    1    2    3    4
+                 |----|----|----|----|
+        forecast
+        start    fx hour
+                 |----|----|----|----|
+        00       01   02   03   04   05
+        01            01   02   03   04
+        02                 01   02   03
+        03                      01   02
+
+        """
+        date = self.start_date
+        data = []
+
+        while date <= self.end_date:
+            self.log.debug('Reading file for date: {}'.format(date))
+            forecast_data = None
+
+            # make sure we get a working file. This allows for six tries,
+            # accounting for the fact that we start at forecast hour 1
+            file_time = date
+            for fx_hr in range(1, self.MAX_FORECAST_HOUR + 1):
+                day_folder, file_name = FileHandler.folder_and_file(
+                    file_time, fx_hr, self.file_type
+                )
+
+                try:
+                    if self.file_type == GribFile.SUFFIX:
+                        base_path = os.path.abspath(self.output_dir)
+                        file = os.path.join(base_path, day_folder, file_name)
+                        if os.path.exists(file):
+                            forecast_data = self.file_loader.load(
+                                file, var_map
+                            )
+                        else:
+                            self.log.error('  No file for {}'.format(file))
+
+                    elif self.file_type == NetCdfFile.SUFFIX:
+                        file = [self.output_dir, day_folder, file_name]
+                        forecast_data = self.file_loader.load(file)
+
+                except Exception as e:
+                    self.log.debug(e)
+                    self.log.debug(
+                        '  Could not load forecast hour {} for date {} '
+                        'successfully'.format(fx_hr, date)
+                    )
+
+                if fx_hr == self.MAX_FORECAST_HOUR + 1:
+                    raise IOError(
```

Comment:
While I think the tests are testing that it cannot find 6 forecast hours (like `test_tries_six_forecast_hours`), this never gets called.

Reply:
It is covered. If you run the tests locally you will see the entry.

If you look at the `test_with_loading_error` you will find an expectation of `8`. It basically tries for all the hours, but  is never successful since it always raise an exception on the `load` call. See the `side_effect` in that test.

Reply:
When I ran coverage (`make coverage`) it showed that this line was never hit. I'm still not that familiar with Mock though.

Reply:
Had a second look and run `make coverage` myself. Need to be more obedient :wink: to your findings and this is a bug and the if statement is wrong.
Pushing a fix.

Reply:
Haha, I just know we use this if statement so it's good to make sure it works!